mava/networks/retention.py

```diff
@@ -0,0 +1,309 @@
+            decay_matrix = mask_agents[None, :, :] * decay_matrix
+
+        return decay_matrix
+
+    def _get_decay_matrix_mask_timestep(self, ts_dones: Array) -> Array:
+        """Generates a mask over the timesteps based on the done status of agents."""
+        # Get the shape of the input: batch size and number of timesteps
+        batch, num_ts = ts_dones.shape
+
+        # Initialize the mask
+        timestep_mask = jnp.zeros((batch, num_ts, num_ts), dtype=bool)
+        all_false = jnp.zeros((batch, num_ts, num_ts), dtype=bool)
+
+        # Iterate over the timesteps and apply the mask
+        for i in range(num_ts):
+            done_this_step = ts_dones[:, i, jnp.newaxis, jnp.newaxis]
+            # Block positions below the current timestep
+            ts_done_xs = all_false.at[:, i:, :].set(done_this_step)
+            # Block positions before the current timestep
+            ts_done_ys = all_false.at[:, :, :i].set(done_this_step)
+
+            # Combine the x and y masks to get the mask for the current timestep.
+            timestep_mask |= ts_done_xs & ts_done_ys
+
+        return ~timestep_mask
+
+    def _get_default_decay_matrix(self, dones: Array) -> Array:
+        """Compute the decay matrix without taking into account the timestep-based masking."""
+        # Get the shape of the input: batch size and number of timesteps
+        batch, num_ts = dones.shape
+
+        # Create the n and m matrices
+        n = jnp.arange(num_ts)[:, jnp.newaxis, ...]
+        m = jnp.arange(num_ts)[jnp.newaxis, ...]
+
+        # Decay based on difference in timestep indices.
+        decay_matrix = (self.decay_kappa ** (n - m)) * (n >= m)
+        # Replace NaN values with 0
+        decay_matrix = jnp.nan_to_num(decay_matrix)
+
+        # Adjust for batch size
+        decay_matrix = jnp.broadcast_to(decay_matrix, (batch, num_ts, num_ts))
+
+        return decay_matrix
+
+    def get_xi(self, dones: Array) -> Array:
+        """Computes a decaying matrix 'xi', which decays over time until the first done signal."""
+        # Get done status for each timestep by slicing out the agent dimension
+        timestep_dones = dones[:, :: self.n_agents]
+        batch, num_ts = timestep_dones.shape
+
+        # Compute the first done step for each sequence,
+        # or set it to sequence length if no dones exist
+        first_dones = jnp.where(
+            ~jnp.any(timestep_dones, axis=1, keepdims=True),
+            jnp.full((batch, 1), num_ts),
+            jnp.argmax(timestep_dones, axis=1, keepdims=True),
+        )
+
+        # Initialize the decaying matrix 'xi'
+        xi = jnp.zeros((batch, num_ts, 1))
+        # Fill 'xi' with decaying values up until the first done step
+        for i in range(num_ts):
+            before_first_done = i < first_dones
+            xi_i = (self.decay_kappa ** (i + 1)) * before_first_done
+            xi = xi.at[:, i, :].set(xi_i)
```

Comment:
Feels like we should be able to do this without a for loop and just a where, should be able to join with the one above, but I'm not 100% sure

Reply:
I will keep this comment unsolved and we back to it in a discussion